core/trino-main/src/main/java/io/trino/sql/gen/DereferenceCodeGenerator.java

```diff
@@ -93,4 +93,13 @@ public BytecodeNode generateExpression(BytecodeGeneratorContext generator)
 
         return block;
     }
+
+    private static int toIntValue(Object value)
+    {
+        if (value instanceof Number number) {
+            return number.intValue();
```

Comment:
Using `Number` interface is strongly discouraged. In the past it led to bugs like unintentional loss of precision (eg when value is double), silent truncation (when value is long).

Be explicit what type you expect here, and do the checked conversion.